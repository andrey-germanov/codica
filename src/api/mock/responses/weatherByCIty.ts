export const mockWeatherByCity = (name: string, lat: number, lon: number) => {
  const mockWeather = {
    name,
    value: {
      coord: {
        lon,
        lat,
      },
      weather: [
        {
          id: 501,
          main: "Rain",
          description: "moderate rain",
          icon: "10d",
        },
      ],
      base: "stations",
      main: {
        temp: 999,
        feels_like: 999,
        temp_min: 999,
        temp_max: 999,
        pressure: 1015,
        humidity: 64,
        sea_level: 1015,
        grnd_level: 933,
      },
      visibility: 10000,
      wind: {
        speed: 0.62,
        deg: 349,
        gust: 1.18,
      },
      rain: {
        "1h": 3.16,
      },
      clouds: {
        all: 100,
      },
      dt: 1661870592,
      sys: {
        type: 2,
        id: 2075663,
        country: "IT",
        sunrise: 1661834187,
        sunset: 1661882248,
      },
      timezone: 7200,
      id: 3163858,
      name: "Zocca",
      cod: 200,
    },
  };
  return mockWeather;
};
