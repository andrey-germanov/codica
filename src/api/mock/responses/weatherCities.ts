export const mockWeatherCities = (cities: string[]) => {
  const mockWeather = {
    name: 'odessa',
    value: {
      coord: {
        lon: 23,
        lat: 23,
      },
      weather: [
        {
          id: 501,
          main: "Rain",
          description: "moderate rain",
          icon: "10d",
        },
      ],
      base: "stations",
      main: {
        temp: 111,
        feels_like: 111,
        temp_min: 0,
        temp_max: 222,
        pressure: 1015,
        humidity: 64,
        sea_level: 1015,
        grnd_level: 933,
      },
      visibility: 10000,
      wind: {
        speed: 0.62,
        deg: 349,
        gust: 1.18,
      },
      rain: {
        "1h": 3.16,
      },
      clouds: {
        all: 100,
      },
      dt: 1661870592,
      sys: {
        type: 2,
        id: 2075663,
        country: "IT",
        sunrise: 1661834187,
        sunset: 1661882248,
      },
      timezone: 7200,
      id: 3163858,
      name: "Zocca",
      cod: 200,
    },
  };
  const mockArray = [];

  for (let i = 0; i < cities.length; i++) {
    mockArray.push({...mockWeather, name: cities[i]})
  }

  return mockArray;
};
