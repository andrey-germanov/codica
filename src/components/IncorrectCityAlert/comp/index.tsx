import { Alert } from "@mui/material";
import React from "react";

type props = {
  incorrectCity: string;
};

export const IncorrectCityAlertComp = ({ incorrectCity }: props) => {
  return (
    <Alert
      severity="error"
      style={{
        position: "fixed",
        zIndex: "2",
      }}
    >
      incorrect city, it will not be added for your list cities:
      <b data-testid='incorrectCity'>{incorrectCity}</b>
    </Alert>
  );
};
