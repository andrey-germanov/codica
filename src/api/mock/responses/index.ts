import { mockWeatherByCity } from './weatherByCIty';
import { mockWeatherCities } from './weatherCities';

export default {
  mockWeatherByCity,
  mockWeatherCities
}