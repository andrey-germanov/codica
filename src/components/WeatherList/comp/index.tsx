import React, { useState } from "react";
import { Link } from "react-router-dom";
import {
  Card,
  CardActions,
  CardContent,
  TextField,
  Typography,
  Button,
} from "@mui/material";
import RefreshIcon from "@mui/icons-material/Refresh";
import { IncorrectCityAlert } from "../../IncorrectCityAlert";
import { IWeatherCity } from "../../../store/typeStore";

interface IProps {
  weather?: IWeatherCity[];
  RefreshWeather: (lat: number, lon: number, name: string) => void;
  RemoveCity: (name: string) => void;
  AddCity: (name: string) => void;
}

export const WeatherListComp = ({
  weather,
  RefreshWeather,
  RemoveCity,
  AddCity,
}: IProps) => {
  const [inputNewCity, setInputNewCity] = useState("");
  const [typeError, setTypeError] = useState(false);

  const addCity = () => {
    if (inputNewCity.length <= 2) return;

    const validation: RegExpMatchArray | null =
      inputNewCity.match(/^[А-Яа-яЁё -]+$/g);
    if (validation !== null) return setTypeError(true);

    setTypeError(false);
    AddCity(inputNewCity); // 1
    setInputNewCity("");
  };
  const handlerInputNewCity = (e: React.ChangeEvent<HTMLInputElement>) =>
    setInputNewCity(e.target.value);

  const handlerRefreshWeather = (
    lat: number,
    lon: number,
    name: string,
    e: React.MouseEvent<SVGSVGElement, MouseEvent>
  ) => {
    e.preventDefault();
    RefreshWeather(lat, lon, name);
  };

  const handlerRemoveCity = (
    name: string,
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    RemoveCity(name);
  };
  return (
    <div id="WeatherListComp">
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          gap: "20px",
          padding: "20px 0",
        }}
      >
        <IncorrectCityAlert />
        <TextField
          error={typeError}
          id="outlined-basic"
          label="Type City"
          variant="outlined"
          value={inputNewCity}
          onChange={handlerInputNewCity}
        />
        <Button id="addCity" variant="contained" size="large" onClick={addCity}>
          add city
        </Button>
        <Typography sx={{ fontSize: 12 }} color="text.secondary" gutterBottom>
          {process.env.REACT_APP_API_MODE === "mock"
            ? "mock data"
            : "real data"}
        </Typography>
      </div>

      <div
        style={{
          display: "flex",
          flexWrap: "wrap",
          gap: "20px",
          justifyContent: "center",
        }}
      >
        {weather && weather.map((city, key) => {
          if (!city.value) return;
          return (
            <Link
              to={`${city.name}`}
              style={{ textDecoration: "none" }}
              key={key}
              id={city.name}
            >
              <Card
                sx={{ minWidth: 275 }}
                style={{
                  backgroundColor: "#99c3f8",
                }}
                key={key}
              >
                <CardContent>
                  <Typography
                    sx={{ fontSize: 18 }}
                    color="text.secondary"
                    gutterBottom
                  >
                    {city.name}
                  </Typography>
                  <Typography variant="h5" component="div">
                    {city.value.weather[0].main}
                  </Typography>
                  <Typography variant="body2">
                    Temperature: {city.value.main.temp}ºС
                  </Typography>
                  <Typography variant="body2">
                    Feels like: {city.value.main.feels_like}ºC
                  </Typography>
                  <img
                    src={`http://openweathermap.org/img/wn/${city.value.weather[0].icon}@2x.png`}
                    alt={city.value.weather[0].description}
                  />
                </CardContent>
                <CardActions
                  style={{
                    display: "flex",
                    justifyContent: "space-between",
                  }}
                >
                  <Button
                    id="button"
                    variant="text"
                    onClick={(e) => handlerRemoveCity(city.name, e)}
                  >
                    delete city
                  </Button>
                  <RefreshIcon
                    onClick={(e) =>
                      handlerRefreshWeather(
                        city.value.coord.lat,
                        city.value.coord.lon,
                        city.name,
                        e
                      )
                    }
                    style={{ cursor: "pointer" }}
                  />
                </CardActions>
              </Card>
            </Link>
          );
        })}
      </div>
    </div>
  );
};
