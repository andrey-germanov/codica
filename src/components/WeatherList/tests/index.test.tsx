import { shallow, ShallowWrapper } from "enzyme";
import { WeatherListComp } from "../comp";
import responses from "../../../api/mock/responses";
import sinon from "sinon";
import toJson from "enzyme-to-json";

describe("should render WeatherListComp component", () => {
  const mockCities = ["Odessa", "Kiev"];

  const props = {
    weather: responses.mockWeatherCities(mockCities),
    RefreshWeather: sinon.spy(),
    RemoveCity: sinon.spy(),
    AddCity: sinon.spy(),
  };

  const comp: ShallowWrapper<
    any,
    Readonly<{}>,
    React.Component<{}, {}, any>
  > = shallow(<WeatherListComp {...props} />);

  expect(toJson(comp)).toMatchSnapshot();
  
  describe("has weather", () => {
    const component = shallow(<WeatherListComp {...props} />);
    it("simple props", () => {
      const links = component.find("Link");

      expect(links.length).toBe(props.weather.length);
    });
  });

  describe("has no weather", () => {
    it("props without weather", () => {
      const props = {
        RefreshWeather: sinon.spy(),
        RemoveCity: sinon.spy(),
        AddCity: sinon.spy(),
      };
      const component = shallow(<WeatherListComp {...props} />);
      const links = component.find("Link");

      expect(links).toHaveLength(0);
    });
  });

  it("should contain .WeatherListComp wrapper", () => {
    const wrapperId = comp.find("#WeatherListComp");

    expect(wrapperId).toHaveLength(1);
  });

  it("should render city card ", () => {
    const cards = comp.find(`#${mockCities[0]}`);
    expect(cards).toHaveLength(1);
  });

  it("should contain links equal count cities in weather array ", () => {
    const links = comp.find("Link");

    expect(links).toHaveLength(props.weather.length);
  });
});
