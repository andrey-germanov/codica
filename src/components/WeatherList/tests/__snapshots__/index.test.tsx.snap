// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  id="WeatherListComp"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "gap": "20px",
        "justifyContent": "center",
        "padding": "20px 0",
      }
    }
  >
    <IncorrectCityAlert />
    <ForwardRef(TextField)
      error={false}
      id="outlined-basic"
      label="Type City"
      onChange={[Function]}
      value=""
      variant="outlined"
    />
    <ForwardRef(Button)
      id="addCity"
      onClick={[Function]}
      size="large"
      variant="contained"
    >
      add city
    </ForwardRef(Button)>
    <ForwardRef(Typography)
      color="text.secondary"
      gutterBottom={true}
      sx={
        Object {
          "fontSize": 12,
        }
      }
    >
      real data
    </ForwardRef(Typography)>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "gap": "20px",
        "justifyContent": "center",
      }
    }
  >
    <Link
      id="Odessa"
      key="0"
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to="Odessa"
    >
      <ForwardRef(Card)
        key="0"
        style={
          Object {
            "backgroundColor": "#99c3f8",
          }
        }
        sx={
          Object {
            "minWidth": 275,
          }
        }
      >
        <ForwardRef(CardContent)>
          <ForwardRef(Typography)
            color="text.secondary"
            gutterBottom={true}
            sx={
              Object {
                "fontSize": 18,
              }
            }
          >
            Odessa
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            component="div"
            variant="h5"
          >
            Rain
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            variant="body2"
          >
            Temperature: 
            111
            ºС
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            variant="body2"
          >
            Feels like: 
            111
            ºC
          </ForwardRef(Typography)>
          <img
            alt="moderate rain"
            src="http://openweathermap.org/img/wn/10d@2x.png"
          />
        </ForwardRef(CardContent)>
        <ForwardRef(CardActions)
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <ForwardRef(Button)
            id="button"
            onClick={[Function]}
            variant="text"
          >
            delete city
          </ForwardRef(Button)>
          <Memo(ForwardRef(RefreshIcon))
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          />
        </ForwardRef(CardActions)>
      </ForwardRef(Card)>
    </Link>
    <Link
      id="Kiev"
      key="1"
      style={
        Object {
          "textDecoration": "none",
        }
      }
      to="Kiev"
    >
      <ForwardRef(Card)
        key="1"
        style={
          Object {
            "backgroundColor": "#99c3f8",
          }
        }
        sx={
          Object {
            "minWidth": 275,
          }
        }
      >
        <ForwardRef(CardContent)>
          <ForwardRef(Typography)
            color="text.secondary"
            gutterBottom={true}
            sx={
              Object {
                "fontSize": 18,
              }
            }
          >
            Kiev
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            component="div"
            variant="h5"
          >
            Rain
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            variant="body2"
          >
            Temperature: 
            111
            ºС
          </ForwardRef(Typography)>
          <ForwardRef(Typography)
            variant="body2"
          >
            Feels like: 
            111
            ºC
          </ForwardRef(Typography)>
          <img
            alt="moderate rain"
            src="http://openweathermap.org/img/wn/10d@2x.png"
          />
        </ForwardRef(CardContent)>
        <ForwardRef(CardActions)
          style={
            Object {
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <ForwardRef(Button)
            id="button"
            onClick={[Function]}
            variant="text"
          >
            delete city
          </ForwardRef(Button)>
          <Memo(ForwardRef(RefreshIcon))
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          />
        </ForwardRef(CardActions)>
      </ForwardRef(Card)>
    </Link>
  </div>
</div>
`;
